File: problems/problemset/url-router/Python/solution/router.py
from dataclasses import dataclass, field

@dataclass
class RouteMatch:
    pattern: str = ''
    params: list[tuple[str, str]] = field(default_factory=list)

class Router:
    def __init__(self):
        self._routes: list[tuple[str, str, list[str]]] = []  # (method, pattern, segments)

    def register(self, method: str, pattern: str) -> None:
        self._routes.append((method, pattern, pattern.split('/')))

    def match(self, method: str, url: str) -> RouteMatch:
        url_segs = url.split('/')
        best_pattern = None
        best_score = None
        best_params = None

        for route_method, pattern, segs in self._routes:
            if route_method != method or len(segs) != len(url_segs):
                continue
            score = []
            params = []
            ok = True
            for ps, us in zip(segs, url_segs):
                if ps.startswith(':'):
                    params.append((ps[1:], us))
                    score.append(1)
                elif ps == us:
                    score.append(0)
                else:
                    ok = False
                    break
            if ok and (best_score is None or score < best_score):
                best_pattern = pattern
                best_score = score
                best_params = params

        if best_pattern is None:
            return RouteMatch()
        return RouteMatch(pattern=best_pattern, params=best_params)

File: problems/problemset/url-router/Rust/solution/router.rs
pub struct RouteMatch {
    pub pattern: String,
    pub params: Vec<(String, String)>,
}

impl RouteMatch {
    fn empty() -> Self {
        Self { pattern: String::new(), params: vec![] }
    }
}

struct Route {
    method: String,
    pattern: String,
    segments: Vec<String>,
}

pub struct Router {
    routes: Vec<Route>,
}

impl Router {
    pub fn new() -> Self {
        Self { routes: vec![] }
    }

    pub fn register(&mut self, method: &str, pattern: &str) {
        self.routes.push(Route {
            method: method.to_string(),
            pattern: pattern.to_string(),
            segments: pattern.split('/').map(|s| s.to_string()).collect(),
        });
    }

    pub fn match_route(&self, method: &str, url: &str) -> RouteMatch {
        let url_segs: Vec<&str> = url.split('/').collect();
        let mut best_pattern: Option<&str> = None;
        let mut best_score: Option<Vec<u8>> = None;
        let mut best_params: Vec<(String, String)> = vec![];

        for route in &self.routes {
            if route.method != method || route.segments.len() != url_segs.len() {
                continue;
            }
            let mut score = vec![];
            let mut params = vec![];
            let mut ok = true;
            for (ps, us) in route.segments.iter().zip(url_segs.iter()) {
                if ps.starts_with(':') {
                    params.push((ps[1..].to_string(), us.to_string()));
                    score.push(1u8);
                } else if ps == us {
                    score.push(0u8);
                } else {
                    ok = false;
                    break;
                }
            }
            if ok && (best_score.is_none() || score < *best_score.as_ref().unwrap()) {
                best_pattern = Some(&route.pattern);
                best_score = Some(score);
                best_params = params;
            }
        }

        match best_pattern {
            Some(p) => RouteMatch { pattern: p.to_string(), params: best_params },
            None => RouteMatch::empty(),
        }
    }
}
